[Feature] Support new feature for backup/restore query and metric (#17526)

srlch opened the pull request:
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #17526

## Problem Summary(Required) ：
add some new feature for backup/restore:
1. support show backup / show restore to show all databases' backup/restore info
2. support show snapshot on repo where in (snapshotname) to show specified snapshot
3. add starrocks_fe_unfinished_backup_job AND starrocks_fe_unfinished_restore_job to indicate the unfinished backup/restore job

```
MySQL [test1]> show backup;
+-------+--------------+--------+----------+------------+---------------------+----------------------+---------------------+---------------------+-----------------+----------+------------+--------+---------+
| JobId | SnapshotName | DbName | State    | BackupObjs | CreateTime          | SnapshotFinishedTime | UploadFinishedTime  | FinishedTime        | UnfinishedTasks | Progress | TaskErrMsg | Status | Timeout |
+-------+--------------+--------+----------+------------+---------------------+----------------------+---------------------+---------------------+-----------------+----------+------------+--------+---------+
| 10114 | mysnapshot   | test   | FINISHED | [test.t]   | 2023-02-17 08:54:52 | 2023-02-17 08:54:57  | 2023-02-17 08:55:03 | 2023-02-17 08:55:09 |                 |          |            | [OK]   | 86400   |
| 10118 | mysnapshot1  | test1  | FINISHED | [test1.t]  | 2023-02-17 08:55:05 | 2023-02-17 08:55:09  | 2023-02-17 08:55:15 | 2023-02-17 08:55:21 |                 |          |            | [OK]   | 86400   |
+-------+--------------+--------+----------+------------+---------------------+----------------------+---------------------+---------------------+-----------------+----------+------------+--------+---------+

MySQL [test1]> show snapshot on test_repo where snapshot in (mysnapshot, mysnapshot1);
+-------------+-------------------------+--------+
| Snapshot    | Timestamp               | Status |
+-------------+-------------------------+--------+
| mysnapshot  | 2023-02-17-08-54-52-012 | OK     |
| mysnapshot1 | 2023-02-17-08-55-05-164 | OK     |
+-------------+-------------------------+--------+

MySQL [test1]> show restore;
+-------+-------------+-------------------------+--------+----------+-----------+----------------+-------------+---------------------+---------------------+----------------------+----------------------+---------------------+-----------------+----------+------------+--------+---------+
| JobId | Label       | Timestamp               | DbName | State    | AllowLoad | ReplicationNum | RestoreObjs | CreateTime          | MetaPreparedTime    | SnapshotFinishedTime | DownloadFinishedTime | FinishedTime        | UnfinishedTasks | Progress | TaskErrMsg | Status | Timeout |
+-------+-------------+-------------------------+--------+----------+-----------+----------------+-------------+---------------------+---------------------+----------------------+----------------------+---------------------+-----------------+----------+------------+--------+---------+
| 10122 | mysnapshot  | 2023-02-17-08-54-52-012 | test   | FINISHED | false     | 1              | t           | 2023-02-17 08:57:28 | 2023-02-17 08:57:30 | 2023-02-17 08:57:33  | 2023-02-17 08:57:39  | 2023-02-17 08:57:45 |                 |          |            | [OK]   | 86400   |
| 10140 | mysnapshot1 | 2023-02-17-08-55-05-164 | test1  | FINISHED | false     | 1              | t           | 2023-02-17 08:57:43 | 2023-02-17 08:57:45 | 2023-02-17 08:57:48  | 2023-02-17 08:57:54  | 2023-02-17 08:58:00 |                 |          |            | [OK]   | 86400   |
+-------+-------------+-------------------------+--------+----------+-----------+----------------+-------------+---------------------+---------------------+----------------------+----------------------+---------------------+-----------------+----------+------------+--------+---------+
```

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

srlch commented:
run starrocks_fe_unittest

chaoyli commented on fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:
```diff
@@ -1360,6 +1360,7 @@ cancelBackupStatement
 
 showBackupStatement
     : SHOW BACKUP ((FROM | IN) identifier)?
+    | SHOW ALL BACKUP
```
What's the difference between `show all backup` and `show backup`?

srlch replied:
show all backup will show all backup info from all different database with backup task. But show backup just show the info of current used database.

chaoyli commented on fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:
```diff
@@ -1375,11 +1376,13 @@ cancelRestoreStatement
 showSnapshotStatement
     : SHOW SNAPSHOT ON identifier
     (WHERE expression)?
+    | SHOW SNAPSHOT ON identifier IN stringList
```
snow snapshot on where?

srlch replied:
where clause don't support multiple snapshotName filtering

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 55 / 84 (65.48%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/backup/RestoreJob.java|1|3|33.33%|[434, 1212]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|24|49|48.98%|[1802, 1803, 1813, 1814, 1815, 1827, 1857, 1858, 1859, 1861, 1862, 1863, 1864, 1866, 1867, 1869, 1872, 1873, 1874, 1875, 1876, 1879, 1880, 1881, 1882]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowSnapshotStmt.java|3|4|75.00%|[48]|
|:large_blue_circle:|com/starrocks/backup/Repository.java|4|5|80.00%|[593]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/BackupRestoreAnalyzer.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/metric/MetricRepo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/backup/BackupJob.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowRestoreStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowBackupStmt.java|3|3|100.00%|[]|

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 67 / 101 (66.34%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/backup/RestoreJob.java|1|3|33.33%|[434, 1212]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|24|49|48.98%|[1802, 1803, 1813, 1814, 1815, 1827, 1857, 1858, 1859, 1861, 1862, 1863, 1864, 1866, 1867, 1869, 1872, 1873, 1874, 1875, 1876, 1879, 1880, 1881, 1882]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowSnapshotAnalyzer.java|12|16|75.00%|[87, 91, 97, 102]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowSnapshotStmt.java|4|5|80.00%|[49]|
|:large_blue_circle:|com/starrocks/backup/Repository.java|4|5|80.00%|[593]|
|:large_blue_circle:|com/starrocks/sql/analyzer/BackupRestoreAnalyzer.java|4|5|80.00%|[269]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/metric/MetricRepo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/backup/BackupJob.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowRestoreStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowBackupStmt.java|3|3|100.00%|[]|

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 69 / 102 (67.65%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/backup/RestoreJob.java|1|3|33.33%|[434, 1212]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|24|49|48.98%|[1803, 1804, 1814, 1815, 1816, 1828, 1858, 1859, 1860, 1862, 1863, 1864, 1865, 1867, 1868, 1870, 1873, 1874, 1875, 1876, 1877, 1880, 1881, 1882, 1883]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowSnapshotAnalyzer.java|12|16|75.00%|[87, 91, 97, 102]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowSnapshotStmt.java|4|5|80.00%|[49]|
|:large_blue_circle:|com/starrocks/backup/Repository.java|4|5|80.00%|[593]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/BackupRestoreAnalyzer.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/metric/MetricRepo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/backup/BackupJob.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowRestoreStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowBackupStmt.java|3|3|100.00%|[]|

srlch committed:
[Feature] Support new feature for backup/restore query and metric (#17526)

add some new feature for backup/restore:
1. support show backup / show restore to show all databases' backup/restore info
2. support show snapshot on repo where in (snapshotname) to show specified snapshot
3. add starrocks_fe_unfinished_backup_job AND starrocks_fe_unfinished_restore_job to indicate the unfinished backup/restore job

eg.
MySQL [test1]> show backup;
+-------+--------------+--------+----------+------------+---------------------+----------------------+---------------------+---------------------+-----------------+----------+------------+--------+---------+
| JobId | SnapshotName | DbName | State    | BackupObjs | CreateTime          | SnapshotFinishedTime | UploadFinishedTime  | FinishedTime        | UnfinishedTasks | Progress | TaskErrMsg | Status | Timeout |
+-------+--------------+--------+----------+------------+---------------------+----------------------+---------------------+---------------------+-----------------+----------+------------+--------+---------+
| 10114 | mysnapshot   | test   | FINISHED | [test.t]   | 2023-02-17 08:54:52 | 2023-02-17 08:54:57  | 2023-02-17 08:55:03 | 2023-02-17 08:55:09 |                 |          |            | [OK]   | 86400   |
| 10118 | mysnapshot1  | test1  | FINISHED | [test1.t]  | 2023-02-17 08:55:05 | 2023-02-17 08:55:09  | 2023-02-17 08:55:15 | 2023-02-17 08:55:21 |                 |          |            | [OK]   | 86400   |
+-------+--------------+--------+----------+------------+---------------------+----------------------+---------------------+---------------------+-----------------+----------+------------+--------+---------+

MySQL [test1]> show snapshot on test_repo where snapshot in (mysnapshot, mysnapshot1);
+-------------+-------------------------+--------+
| Snapshot    | Timestamp               | Status |
+-------------+-------------------------+--------+
| mysnapshot  | 2023-02-17-08-54-52-012 | OK     |
| mysnapshot1 | 2023-02-17-08-55-05-164 | OK     |
+-------------+-------------------------+--------+

MySQL [test1]> show restore;
+-------+-------------+-------------------------+--------+----------+-----------+----------------+-------------+---------------------+---------------------+----------------------+----------------------+---------------------+-----------------+----------+------------+--------+---------+
| JobId | Label       | Timestamp               | DbName | State    | AllowLoad | ReplicationNum | RestoreObjs | CreateTime          | MetaPreparedTime    | SnapshotFinishedTime | DownloadFinishedTime | FinishedTime        | UnfinishedTasks | Progress | TaskErrMsg | Status | Timeout |
+-------+-------------+-------------------------+--------+----------+-----------+----------------+-------------+---------------------+---------------------+----------------------+----------------------+---------------------+-----------------+----------+------------+--------+---------+
| 10122 | mysnapshot  | 2023-02-17-08-54-52-012 | test   | FINISHED | false     | 1              | t           | 2023-02-17 08:57:28 | 2023-02-17 08:57:30 | 2023-02-17 08:57:33  | 2023-02-17 08:57:39  | 2023-02-17 08:57:45 |                 |          |            | [OK]   | 86400   |
| 10140 | mysnapshot1 | 2023-02-17-08-55-05-164 | test1  | FINISHED | false     | 1              | t           | 2023-02-17 08:57:43 | 2023-02-17 08:57:45 | 2023-02-17 08:57:48  | 2023-02-17 08:57:54  | 2023-02-17 08:58:00 |                 |          |            | [OK]   | 86400   |
+-------+-------------+-------------------------+--------+----------+-----------+----------------+-------------+---------------------+---------------------+----------------------+----------------------+---------------------+-----------------+----------+------------+--------+---------+

Signed-off-by: srlch <linzichao@starrocks.com>

srlch commented:
run starrocks_fe_unittest

srlch committed:
fix fe case fail

Signed-off-by: srlch <linzichao@starrocks.com>

srlch commented:
run starrocks_fe_unittest

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 50 / 73 (68.49%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/backup/RestoreJob.java|1|3|33.33%|[434, 1219]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|22|37|59.46%|[1802, 1803, 1850, 1851, 1853, 1854, 1857, 1858, 1859, 1860, 1861, 1864, 1865, 1866, 1867]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowSnapshotAnalyzer.java|12|16|75.00%|[87, 91, 97, 102]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowSnapshotStmt.java|4|5|80.00%|[49]|
|:large_blue_circle:|com/starrocks/backup/Repository.java|4|5|80.00%|[593]|
|:large_blue_circle:|com/starrocks/metric/MetricRepo.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/backup/BackupJob.java|3|3|100.00%|[]|

wanpengfei-git commented:
run starrocks_admit_test
